source/agora/network/RPC.d

```diff
@@ -0,0 +1,468 @@
+                   __FILE__.ptr, __LINE__);
+        }
+        //stream.close();
+    }
+}
+
+/*******************************************************************************
+
+    Handle incoming (server) requests
+
+    This function will read from `stream`, find which function of `api` to call,
+    deserialize the parameters accordingly, then call `api`.
+    If anything goes wrong (deserialization failed, wrong method, etc...),
+    a trace message will be logged and `api` won't be called.
+
+    Params:
+      API = The type of API that will handle the call
+      api = The object that will handle the call
+      stream = The TCP stream to read data from
+
+*******************************************************************************/
+
+private void handleThrow (API) (scope API api, ref TCPConnection stream)
+    @trusted
+{
+    static string[Hash] lookup;
+    if (lookup.length == 0)
+    {
+        static foreach (member; __traits(allMembers, API))
+            static foreach (ovrld; __traits(getOverloads, API, member))
+                lookup[hashFull(ovrld.mangleof)] = ovrld.mangleof;
+    }
+
+    log.info("[{}] Handling a new request: {}", stream.peerAddress, stream.leastSize());
+    Hash methodbin;
+    stream.read(methodbin[]);
+    const method = methodbin in lookup;
+    if (method is null)
+    {
+        log.error("[{}] Calling out of range method: {}",
+                  stream.peerAddress, methodbin);
+        return;
+    }
+
+    ubyte[1024] buffer = void;
+    scope DeserializeDg reader = (size_t size) @safe
+    {
+        assert(size < buffer.length);
+        scope (failure) assert(0);
+        stream.read(buffer[0 .. size]);
+        return buffer[0 .. size];
+    };
+
+    // Helper template for staticMap
+    Target convert (Target) ()
+    {
+        return deserializeFull!Target(reader);
+    }
+
+    switch (*method)
+    {
+    static foreach (member; __traits(allMembers, API))
+        static foreach (ovrld; __traits(getOverloads, API, member))
+        {
+        case ovrld.mangleof:
+            enum CallMixin = "api." ~ member ~ "(staticMap!(convert, Parameters!ovrld));";
+
+            log.trace("[SERVER] {} requested {}({})",
+                      stream.peerAddress, member, (Parameters!ovrld).stringof);
+
+            // Call functions + return
+            static if (is(ReturnType!ovrld == void))
+            {
+                mixin(CallMixin);
+                log.trace("[SERVER] Goodbye {}", methodbin);
+                stream.write(methodbin[]);
+            }
+            else
+            {
+                mixin("auto foo = ", CallMixin);
+                log.trace("[SERVER] Returning {}", foo);
+                serializePart(foo, (in ubyte[] v) { stream.write(v); });
+                log.trace("[SERVER] Done writing...");
+            }
+            return;
+        }
+    default:
+        assert(0);
+    }
+}
+
+// /// Generate an enum used for the dispatch of functions
```

Comment:
Are these implementations for the future?

Reply:
Those should be removed.